src/download.rs

```diff
@@ -295,4 +318,346 @@ mod tests {
 
         Ok(())
     }
+    /// Tests that pre message is sent for attachment larger than `PRE_MSG_ATTACHMENT_SIZE_THRESHOLD`
+    /// Also test that pre message is sent first, before the full message
+    /// And that Autocrypt-gossip and selfavatar never go into full-messages
+    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+    async fn test_sending_pre_message() -> Result<()> {
+        let mut tcm = TestContextManager::new();
+        let alice = tcm.alice().await;
+        let bob = tcm.bob().await;
+        let fiona = tcm.fiona().await;
+        let group_id = alice
+            .create_group_with_members("test group", &[&bob, &fiona])
+            .await;
+
+        let mut msg = Message::new(Viewtype::File);
+        msg.set_file_from_bytes(&alice.ctx, "test.bin", &[0u8; 300_000], None)?;
+        msg.set_text("test".to_owned());
+
+        // assert that test attachment is bigger than limit
+        assert!(msg.get_filebytes(&alice.ctx).await?.unwrap() > PRE_MSG_ATTACHMENT_SIZE_THRESHOLD);
+
+        let msg_id = chat::send_msg(&alice.ctx, group_id, &mut msg).await?;
+        let smtp_rows = alice.get_smtp_rows_for_msg(msg_id).await;
+
+        //   pre-message and full message should be present
+        //   and test that correct headers are present on both messages
+        assert_eq!(smtp_rows.len(), 2);
+        let pre_message_bytes = smtp_rows
+            .first()
+            .expect("first element exists")
+            .2
+            .as_bytes();
+        let pre_message = mailparse::parse_mail(pre_message_bytes)?;
+        let full_message_bytes = smtp_rows
+            .get(1)
+            .expect("second element exists")
+            .2
+            .as_bytes();
```

Comment:
Since there are multiple tests that need both a full message and a pre-message, would be good to have a function that:
- uses `pop_sent_msg()` to pop the full-message
- uses `pop_sent_msg()` to pop the pre-message
- uses `pop_sent_msg_opt()` to check that no more than these two messages were sent
- returns the pre- and full-message

`get_smtp_rows_for_msg()` is probably not needed, then, which is also nice; returning a tuple and then accessing one element `.2` is hard to read.